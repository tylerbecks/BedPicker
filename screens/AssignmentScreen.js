import React from "react";
import { StyleSheet, View, ScrollView, Text } from "react-native";
import BedSection from "../components/BedSection";
import AvatarItem from "../components/AvatarItem";
import Assigner from "../utils/Assigner";
import Bed from "../utils/classes/Bed";

const getTahoeBeds = () => [
  new Bed("queen", "Master Queen"),
  new Bed("twin", "Upstairs twin 1"),
  new Bed("twin", "Upstairs twin 2"),
  new Bed("full", "Upstairs full")
];

const styles = StyleSheet.create({
  container: {
    backgroundColor: "#fff",
    flex: 1,
    padding: 8
  },
  contentContainer: {},
  date: {
    marginTop: 20
  },
  dateContainer: {
    alignItems: "center"
  },
  sectionHeader: {
    fontWeight: "bold",
    fontSize: 18,
    marginBottom: 12
  }
});

export default class AssignmentScreen extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      assignment: null
    };

    this.sleepers = null;
  }

  componentDidMount() {
    const { navigation } = this.props;

    const sleepers = navigation.getParam("sleepers", []);
    this.setAssignment(sleepers);
  }

  componentWillReceiveProps(nextProps) {
    const sleepers = nextProps.navigation.getParam("sleepers", []);
    if (sleepers !== this.sleepers) {
      this.setAssignment(sleepers);
    }
  }

  setAssignment(sleepers) {
    this.sleepers = sleepers;
    const assigner = new Assigner(sleepers, getTahoeBeds());
    this.setState({ assignment: assigner.createAssignment() });
  }

  getFormattedDate() {
    const { assignment } = this.state;
    if (!assignment || !assignment.date) {
      throw Error("expected assignment date to exist!");
    }

    const date = new Date(assignment.date);

    return date.toLocaleDateString("en-US", {
      weekday: "short",
      year: "numeric",
      month: "short",
      day: "numeric",
      hour: "numeric",
      minute: "numeric"
    });
  }

  render() {
    const { assignment } = this.state;

    if (!assignment) {
      return null;
    }

    return (
      <View style={styles.container}>
        <ScrollView
          style={styles.container}
          contentContainerStyle={styles.contentContainer}
        >
          {assignment.beds.map(({ name, guests }) => (
            <BedSection key={name} name={name} guests={guests} />
          ))}
          <Text style={styles.sectionHeader}>Guests without beds</Text>
          {assignment.guestsWithoutBeds.map(guest => (
            <AvatarItem key={guest.id} text={guest.name} photo={guest.photo} />
          ))}
          <View style={styles.dateContainer}>
            <Text style={styles.date}>{this.getFormattedDate()}</Text>
          </View>
        </ScrollView>
      </View>
    );
  }
}
