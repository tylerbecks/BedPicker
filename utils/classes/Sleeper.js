export default class Sleeper {
  constructor(...guests) {
    this.guests = guests;
    this.nextAssignedIndex = 0;
  }

  getGuest() {
    if (this.allAsigned()) {
      throw new Error("All guests already assigned");
    }
    const guest = this.guests[this.nextAssignedIndex];
    this.nextAssignedIndex += 1;

    return guest;
  }

  allAsigned() {
    return this.nextAssignedIndex >= this.guests.length;
  }

  getUnassignedGuestCount() {
    return this.guests.length - this.nextAssignedIndex;
  }
}
