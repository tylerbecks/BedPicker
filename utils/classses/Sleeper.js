export default class Sleeper {
  constructor(...guests) {
    this.guests = guests;
    this.nextAssignedIndex = 0;
  }

  getGuest = () => {
    const guest = this.guests[this.nextAssignedIndex];
    this.nextAssignedIndex++;
    return guest;
  };

  allAsigned = () => this.nextAssignedIndex >= this.guests.length;
}
