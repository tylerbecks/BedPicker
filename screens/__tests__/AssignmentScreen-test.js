import "react-native";
import React from "react";
import MockDate from "mockdate";
import AssignmentScreen from "../AssignmentScreen";
import renderer from "react-test-renderer";
import Sleeper from "../../utils/classes/Sleeper";

it("renders", () => {
  MockDate.set(1434319925275);

  const getParamMock = jest.fn();
  const sleeper1 = new Sleeper({ id: 1 });
  const sleeper2 = new Sleeper({ id: 2 });
  getParamMock.mockReturnValue([sleeper1, sleeper2]);

  expect(
    renderer
      .create(
        <AssignmentScreen
          navigation={{
            getParam: getParamMock
          }}
        />
      )
      .toJSON()
  ).toMatchSnapshot();

  MockDate.reset();
});
