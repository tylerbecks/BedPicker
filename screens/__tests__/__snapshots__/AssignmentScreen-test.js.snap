// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "padding": 8,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={Object {}}
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "padding": 8,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 12,
            }
          }
        >
          Master Queen
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginLeft": 20,
              "paddingVertical": 8,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "borderRadius": 20,
                "height": 40,
                "marginRight": 10,
                "width": 40,
              }
            }
          />
          <Text
            style={
              Object {
                "fontSize": 14,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 12,
            }
          }
        >
          Upstairs full
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginLeft": 20,
              "paddingVertical": 8,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "borderRadius": 20,
                "height": 40,
                "marginRight": 10,
                "width": 40,
              }
            }
          />
          <Text
            style={
              Object {
                "fontSize": 14,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 12,
            }
          }
        >
          Upstairs twin 1
        </Text>
      </View>
      <View
        style={
          Object {
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 12,
            }
          }
        >
          Upstairs twin 2
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 12,
          }
        }
      >
        Guests without beds
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          Sun, Jun 14, 2015, 3:12 PM
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
